DispatchingAuthenticator gets mixed up under high concurrency

When used in an application with many concurrent requests, I get the warning "previous digest authentication with same nonce failed, returning null" frequently right before I get a failed request.  I'm wondering if somehow the shared "parameters" object in DigestAuthenticator is getting overwritten by multiple concurrent requests and thus accidentally getting marked as stale.  I've been using DispatchingAuthenticator as directed in the README.md, but only using BasicAuthenticator or DigestAuthenticator because of #32.

Ok, you're right.  I knew those would distract so I tried to loosen all those match assertions, but it looks like I missed two.  I've pushed an update so those won't distract.  I believe the following error is not a test bug.  In other words, the behavior of DigestAuthenticator is different when run in a multi-threaded env vs single-threaded env.  Here's what I'm seeing on STDERR:

```
Jul 11, 2017 9:26:47 AM okhttp3.internal.platform.Platform log
WARNING: previous digest authentication with same nonce failed, returning null
Test testAuthenticate_withWrongPassword_shouldNotRepeat failed under testMultithreadedAuthenticate:
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest$1.run(DigestAuthenticatorTest.java:309)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: 
Expected: is null
     but: was <Request{method=GET, url=http://www.google.com/, tag=Request{method=GET, url=http://www.google.com/, tag=null}}>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest.testAuthenticate_withWrongPassword_shouldNotRepeat(DigestAuthenticatorTest.java:216)
	... 8 more
```

I get similar errors sometimes from testAuthenticate__withProxy__shouldWork and testAuthenticate_withDifferentNonce_shouldNotRetry.  I don't always see the "WARNING: previous digest authentication with same nonce failed, returning null" although I was seeing that pretty consistently in the output from my failing stress test.

In case it helps, here are two examples of when the others fail:

```
Test testAuthenticate_withDifferentNonce_shouldNotRetry failed under testMultithreadedAuthenticate:
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest$1.run(DigestAuthenticatorTest.java:309)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: 
Expected: is null
     but: was <Request{method=GET, url=http://www.google.com/, tag=Request{method=GET, url=http://www.google.com/, tag=null}}>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest.testAuthenticate_withDifferentNonce_shouldNotRetry(DigestAuthenticatorTest.java:241)
	... 8 more
Test testAuthenticate__withProxy__shouldWork failed under testMultithreadedAuthenticate:
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest$1.run(DigestAuthenticatorTest.java:309)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: 
Expected: a string matching the pattern 'Digest username="user1", realm="myrealm", nonce="NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4", uri="/", response="[0-9a-f]+", qop=auth, nc=00000001, cnonce="[0-9a-f]+", algorithm=MD5'
     but: was "Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", response=\"0222dac12825f93e1697e8cb34cb5ece\", qop=auth, nc=00000002, cnonce=\"0f9866728b5cd1fc\", algorithm=MD5"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.burgstaller.okhttp.digest.DigestAuthenticatorTest.testAuthenticate__withProxy__shouldWork(DigestAuthenticatorTest.java:98)
	... 8 more
```

--- a/build.gradle
+++ b/build.gradle
@@ -9,12 +9,14 @@ plugins {
 }
 
 apply plugin: 'java'
+apply plugin: 'maven'
 
 if (new File("./gradle/localrepo.gradle").exists()) {
     apply from: "./gradle/localrepo.gradle"
 }
 
 group = "com.burgstaller"
+version = "1.12.1"
 archivesBaseName = 'okhttp-digest'
 
 sourceCompatibility = JavaVersion.VERSION_1_7

--- a/src/main/java/com/burgstaller/okhttp/digest/DigestAuthenticator.java
+++ b/src/main/java/com/burgstaller/okhttp/digest/DigestAuthenticator.java
@@ -18,6 +18,7 @@ import java.security.MessageDigest;
 import java.security.SecureRandom;
 import java.util.ArrayList;
 import java.util.Formatter;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Locale;
@@ -25,6 +26,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.StringTokenizer;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
 
 import okhttp3.Headers;
 import okhttp3.Request;
@@ -61,6 +63,7 @@ public class DigestAuthenticator implements CachingAuthenticator {
             'e', 'f'
     };
 
+    AtomicReference<ConcurrentHashMap<String,String>> parametersRef = new AtomicReference<>();
     private Charset credentialsCharset = Charset.forName("ASCII");
     private final Credentials credentials;
     private String lastNonce;
@@ -138,10 +141,12 @@ public class DigestAuthenticator implements CachingAuthenticator {
     @Override
     public synchronized Request authenticate(Route route, Response response) throws IOException {
         String header = findDigestHeader(response.headers(), getHeaderName(response.code()));
-        Map<String,String> parameters = new ConcurrentHashMap<String,String>();
+        ConcurrentHashMap<String,String> parameters = new ConcurrentHashMap<>();
         parseChallenge(header, 7, header.length() - 7, parameters);
         // first copy all request headers to our params array
         copyHeaderMap(response.headers(), parameters);
+        // save these parameters so future requests don't need the challenge response every time
+        parametersRef.set(parameters);
 
         // sanity check for issue #22
         if (parameters.get("nonce") == null) {
@@ -175,7 +180,9 @@ public class DigestAuthenticator implements CachingAuthenticator {
 
     @Override
     public Request authenticateWithState(Route route, Request request) throws IOException {
-      return authenticateWithState(route, request, new ConcurrentHashMap<String,String>());
+      // make sure we don't modify the values in shared parametersRef instance
+      Map<String,String> parameters = new HashMap<>(parametersRef.get());
+      return authenticateWithState(route, request, parameters);
     }
 
     public Request authenticateWithState(Route route, Request request, Map<String,String> parameters) throws IOException {

--- a/src/main/java/com/burgstaller/okhttp/digest/DigestAuthenticator.java
+++ b/src/main/java/com/burgstaller/okhttp/digest/DigestAuthenticator.java
@@ -61,7 +61,6 @@ public class DigestAuthenticator implements CachingAuthenticator {
             'e', 'f'
     };
 
-    Map<String, String> parameters = new ConcurrentHashMap<>();
     private Charset credentialsCharset = Charset.forName("ASCII");
     private final Credentials credentials;
     private String lastNonce;
@@ -139,6 +138,7 @@ public class DigestAuthenticator implements CachingAuthenticator {
     @Override
     public synchronized Request authenticate(Route route, Response response) throws IOException {
         String header = findDigestHeader(response.headers(), getHeaderName(response.code()));
+        Map<String,String> parameters = new ConcurrentHashMap<String,String>();
         parseChallenge(header, 7, header.length() - 7, parameters);
         // first copy all request headers to our params array
         copyHeaderMap(response.headers(), parameters);
@@ -148,7 +148,7 @@ public class DigestAuthenticator implements CachingAuthenticator {
             throw new IllegalArgumentException("missing nonce in challenge header: " + header);
         }
 
-        return authenticateWithState(route, response.request());
+        return authenticateWithState(route, response.request(), parameters);
     }
 
     private String getHeaderName(int httpStatus) {
@@ -175,13 +175,17 @@ public class DigestAuthenticator implements CachingAuthenticator {
 
     @Override
     public Request authenticateWithState(Route route, Request request) throws IOException {
+      return authenticateWithState(route, request, new ConcurrentHashMap<String,String>());
+    }
+
+    public Request authenticateWithState(Route route, Request request, Map<String,String> parameters) throws IOException {
         final String realm = parameters.get("realm");
         if (realm == null) {
             // missing realm, this would mean that the authenticator is not initialized for this
             // request. (e.g. if you configured the DispatchingAuthenticator.
             return null;
         }
-        final String nonce = getParameter("nonce");
+        final String nonce = parameters.get("nonce");
         if (nonce == null) {
             throw new IllegalArgumentException("missing nonce in challenge");
         }
@@ -198,21 +202,21 @@ public class DigestAuthenticator implements CachingAuthenticator {
         if (route == null || !route.requiresTunnel()) {
             final String method = request.method();
             final String uri = RequestLine.requestPath(request.url());
-            getParameters().put("methodname", method);
-            getParameters().put("uri", uri);
+            parameters.put("methodname", method);
+            parameters.put("uri", uri);
         } else {
             final String method = "CONNECT";
             final String uri = request.url().host() + ':' + request.url().port();
-            getParameters().put("methodname", method);
-            getParameters().put("uri", uri);
+            parameters.put("methodname", method);
+            parameters.put("uri", uri);
         }
 
-        final String charset = getParameter("charset");
+        final String charset = parameters.get("charset");
         if (charset == null) {
             String credentialsCharset = getCredentialsCharset(request);
-            getParameters().put("charset", credentialsCharset);
+            parameters.put("charset", credentialsCharset);
         }
-        final NameValuePair digestHeader = createDigestHeader(credentials, request);
+        final NameValuePair digestHeader = createDigestHeader(credentials, request, parameters);
         return request.newBuilder()
                 .header(digestHeader.getName(), digestHeader.getValue())
                 .build();
@@ -245,10 +249,6 @@ public class DigestAuthenticator implements CachingAuthenticator {
         }
     }
 
-    private String getParameter(String key) {
-        return parameters.get(key);
-    }
-
     /**
      * Creates digest-response header as defined in RFC2617.
      *
@@ -258,13 +258,14 @@ public class DigestAuthenticator implements CachingAuthenticator {
 //    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings("LSC_LITERAL_STRING_COMPARISON")
     private synchronized NameValuePair createDigestHeader(
             final Credentials credentials,
-            final Request request) throws AuthenticationException {
-        final String uri = getParameter("uri");
-        final String realm = getParameter("realm");
-        final String nonce = getParameter("nonce");
-        final String opaque = getParameter("opaque");
-        final String method = getParameter("methodname");
-        String algorithm = getParameter("algorithm");
+            final Request request,
+            final Map<String,String> parameters) throws AuthenticationException {
+        final String uri = parameters.get("uri");
+        final String realm = parameters.get("realm");
+        final String nonce = parameters.get("nonce");
+        final String opaque = parameters.get("opaque");
+        final String method = parameters.get("methodname");
+        String algorithm = parameters.get("algorithm");
         // If an algorithm is not specified, default to MD5.
         if (algorithm == null) {
             algorithm = "MD5";
@@ -272,7 +273,7 @@ public class DigestAuthenticator implements CachingAuthenticator {
 
         final Set<String> qopset = new HashSet<>(8);
         int qop = QOP_UNKNOWN;
-        final String qoplist = getParameter("qop");
+        final String qoplist = parameters.get("qop");
         if (qoplist != null) {
             final StringTokenizer tok = new StringTokenizer(qoplist, ",");
             while (tok.hasMoreTokens()) {
@@ -292,7 +293,7 @@ public class DigestAuthenticator implements CachingAuthenticator {
             throw new AuthenticationException("None of the qop methods is supported: " + qoplist);
         }
 
-        String charset = getParameter("charset");
+        String charset = parameters.get("charset");
         if (charset == null) {
             charset = "ISO-8859-1";
         }
@@ -441,11 +442,6 @@ public class DigestAuthenticator implements CachingAuthenticator {
             BasicHeaderValueFormatter.DEFAULT.formatNameValuePair(buffer, param, !noQuotes);
         }
         return new BasicNameValuePair(headerKey, buffer.toString());
-    }
-
-
-    public Map<String, String> getParameters() {
-        return parameters;
     }
 
 

--- a/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
+++ b/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
@@ -98,7 +98,7 @@ public class DigestAuthenticatorTest {
         assertThat(authenticated.header("Proxy-Authorization"),
                 matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
                   "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", " +
-                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test(expected = IllegalArgumentException.class)
@@ -265,7 +265,7 @@ public class DigestAuthenticatorTest {
         // then
         assertThat(authenticated.header("Authorization"),
                 matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"BBBBBB\", " +
-                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test

--- a/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
+++ b/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
@@ -2,8 +2,7 @@ package com.burgstaller.okhttp.digest;
 
 import org.junit.Before;
 import org.junit.Test;
-import org.mockito.Mock;
-import org.mockito.MockitoAnnotations;
+import static org.mockito.Mockito.mock;
 
 import java.io.IOException;
 import java.lang.reflect.Method;
@@ -37,24 +36,17 @@ import static org.mockito.BDDMockito.given;
 
 public class DigestAuthenticatorTest {
 
-    @Mock
-    private Connection mockConnection;
-    @Mock
-    private Dns mockDns;
-    @Mock
-    private SocketFactory socketFactory;
-    @Mock
-    private Authenticator proxyAuthenticator;
-    @Mock
-    private ProxySelector proxySelector;
-    @Mock
-    Proxy proxy;
     private Route mockRoute;
     private DigestAuthenticator authenticator;
 
     @Before
     public void beforeMethod() {
-        MockitoAnnotations.initMocks(this);
+        Connection mockConnection = mock(Connection.class);
+        Dns mockDns = mock(Dns.class);
+        SocketFactory socketFactory = mock(SocketFactory.class);
+        Authenticator proxyAuthenticator = mock(Authenticator.class);
+        ProxySelector proxySelector = mock(ProxySelector.class);
+        Proxy proxy = mock(Proxy.class);
 
         // setup some dummy data so that we dont get NPEs
         Address address = new Address("localhost", 8080, mockDns, socketFactory, null, null,
@@ -305,35 +297,35 @@ public class DigestAuthenticatorTest {
 
     @Test
     public void testMultithreadedAuthenticate() throws Exception {
-      final DigestAuthenticatorTest test = this;
-      final ConcurrentHashMap<String,Exception> exceptions = new ConcurrentHashMap<>();
-      ExecutorService executor = new ThreadPoolExecutor(20, 20, 1, TimeUnit.MINUTES, new LinkedBlockingQueue());
-      for ( final Method method : this.getClass().getDeclaredMethods() ) {
-        if ( method.getName().startsWith("testAuthenticate") ) {
-          executor.execute(new Runnable() {
-            public void run() {
-              try {
-                System.out.println("invoking method=" + method.getName());
-                method.invoke(test);
-              } catch(Exception e) {
-                exceptions.put(method.getName(), e);
-              }
+        final DigestAuthenticatorTest test = this;
+        final ConcurrentHashMap<String,Exception> exceptions = new ConcurrentHashMap<>();
+        ExecutorService executor = new ThreadPoolExecutor(20, 20, 1, TimeUnit.MINUTES, new LinkedBlockingQueue());
+        for ( final Method method : this.getClass().getDeclaredMethods() ) {
+            if ( method.getName().startsWith("testAuthenticate") ) {
+                executor.execute(new Runnable() {
+                    public void run() {
+                        try {
+                            System.out.println("invoking method=" + method.getName());
+                            method.invoke(test);
+                        } catch(Exception e) {
+                            exceptions.put(method.getName(), e);
+                        }
+                    }
+                });
             }
-          });
         }
-      }
-      executor.shutdown();
-      try {
-        executor.awaitTermination(2, TimeUnit.MINUTES);
-      } catch(InterruptedException e) {
-        throw new IllegalStateException("Timeout trying to run testMultithreadedAuthenticate");
-      }
-      if ( exceptions.size() > 0 ) {
-        for ( String methodName : exceptions.keySet() ) {
-          System.err.println("Test " + methodName + " failed under testMultithreadedAuthenticate:");
-          exceptions.get(methodName).printStackTrace();
+        executor.shutdown();
+        try {
+            executor.awaitTermination(2, TimeUnit.MINUTES);
+        } catch(InterruptedException e) {
+            throw new IllegalStateException("Timeout trying to run testMultithreadedAuthenticate");
         }
-        throw exceptions.elements().nextElement();
-      }
+        if ( exceptions.size() > 0 ) {
+            for ( String methodName : exceptions.keySet() ) {
+                System.err.println("Test " + methodName + " failed under testMultithreadedAuthenticate:");
+                exceptions.get(methodName).printStackTrace();
+            }
+            throw exceptions.elements().nextElement();
+        }
     }
 }

--- a/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
+++ b/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
@@ -301,7 +301,8 @@ public class DigestAuthenticatorTest {
         final ConcurrentHashMap<String,Exception> exceptions = new ConcurrentHashMap<>();
         ExecutorService executor = new ThreadPoolExecutor(20, 20, 1, TimeUnit.MINUTES, new LinkedBlockingQueue());
         for ( final Method method : this.getClass().getDeclaredMethods() ) {
-            if ( method.getName().startsWith("testAuthenticate") ) {
+            if ( method.getName().startsWith("testAuthenticate") &&
+                 ! method.getName().endsWith("shouldThrowException") ) {
                 executor.execute(new Runnable() {
                     public void run() {
                         try {

--- a/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
+++ b/src/test/java/com/burgstaller/okhttp/digest/DigestAuthenticatorTest.java
@@ -6,10 +6,16 @@ import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 
 import java.io.IOException;
+import java.lang.reflect.Method;
 import java.net.InetSocketAddress;
 import java.net.Proxy;
 import java.net.ProxySelector;
 import java.util.Collections;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
 import javax.net.SocketFactory;
 
@@ -77,7 +83,9 @@ public class DigestAuthenticatorTest {
         Request authenticated = authenticator.authenticate(mockRoute, response);
 
         assertThat(authenticated.header("Authorization"),
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test
@@ -96,7 +104,9 @@ public class DigestAuthenticatorTest {
         Request authenticated = authenticator.authenticate(mockRoute, response);
 
         assertThat(authenticated.header("Proxy-Authorization"),
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test(expected = IllegalArgumentException.class)
@@ -137,7 +147,10 @@ public class DigestAuthenticatorTest {
 
         String authHeader = authenticated.header("Authorization");
         assertThat(authHeader,
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/path/to/resource\\?parameter=value&parameter2=value2\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", " +
+                  "uri=\"/path/to/resource\\?parameter=value&parameter2=value2\", " +
+                  "response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test
@@ -157,7 +170,10 @@ public class DigestAuthenticatorTest {
 
         String authHeader = authenticated.header("Authorization");
         assertThat(authHeader,
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/path/to/resource\\?parameter=value&parameter2=value2\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", " +
+                  "uri=\"/path/to/resource\\?parameter=value&parameter2=value2\", " +
+                  "response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test
@@ -177,7 +193,9 @@ public class DigestAuthenticatorTest {
         Request authenticated = authenticator.authenticate(mockRoute, response);
 
         assertThat(authenticated.header("Authorization"),
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", " +
+                  "response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test
@@ -185,7 +203,9 @@ public class DigestAuthenticatorTest {
         // given
         Request dummyRequest = new Request.Builder()
                 .url("http://www.google.com")
-                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
+                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", " +
+                  "nonce=\"NnjGCdMhBQA=8ede771f94b593e46e5d0dd10b68313226c133f4\", uri=\"/\", " +
+                  "response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
                 .get()
                 .build();
         Response response = new Response.Builder()
@@ -209,7 +229,8 @@ public class DigestAuthenticatorTest {
         // given
         Request dummyRequest = new Request.Builder()
                 .url("http://www.google.com")
-                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
+                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
                 .get()
                 .build();
         Response response = new Response.Builder()
@@ -233,7 +254,8 @@ public class DigestAuthenticatorTest {
         // given
         Request dummyRequest = new Request.Builder()
                 .url("http://www.google.com")
-                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
+                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
                 .get()
                 .build();
         Response response = new Response.Builder()
@@ -250,7 +272,8 @@ public class DigestAuthenticatorTest {
 
         // then
         assertThat(authenticated.header("Authorization"),
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"BBBBBB\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"BBBBBB\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
     }
 
     @Test
@@ -258,7 +281,8 @@ public class DigestAuthenticatorTest {
         // given
         Request dummyRequest = new Request.Builder()
                 .url("http://www.google.com")
-                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
+                .header("Authorization", "Digest username=\"user1\", realm=\"myrealm\", nonce=\"AAAAAAA\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5")
                 .get()
                 .build();
         Response response = new Response.Builder()
@@ -275,6 +299,41 @@ public class DigestAuthenticatorTest {
 
         // then
         assertThat(authenticated.header("Authorization"),
-                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"BBBBBB\", uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=00000001, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+                matchesPattern("Digest username=\"user1\", realm=\"myrealm\", nonce=\"BBBBBB\", " +
+                  "uri=\"/\", response=\"[0-9a-f]+\", qop=auth, nc=000000\\d\\d, cnonce=\"[0-9a-f]+\", algorithm=MD5"));
+    }
+
+    @Test
+    public void testMultithreadedAuthenticate() throws Exception {
+      final DigestAuthenticatorTest test = this;
+      final ConcurrentHashMap<String,Exception> exceptions = new ConcurrentHashMap<>();
+      ExecutorService executor = new ThreadPoolExecutor(20, 20, 1, TimeUnit.MINUTES, new LinkedBlockingQueue());
+      for ( final Method method : this.getClass().getDeclaredMethods() ) {
+        if ( method.getName().startsWith("testAuthenticate") ) {
+          executor.execute(new Runnable() {
+            public void run() {
+              try {
+                System.out.println("invoking method=" + method.getName());
+                method.invoke(test);
+              } catch(Exception e) {
+                exceptions.put(method.getName(), e);
+              }
+            }
+          });
+        }
+      }
+      executor.shutdown();
+      try {
+        executor.awaitTermination(2, TimeUnit.MINUTES);
+      } catch(InterruptedException e) {
+        throw new IllegalStateException("Timeout trying to run testMultithreadedAuthenticate");
+      }
+      if ( exceptions.size() > 0 ) {
+        for ( String methodName : exceptions.keySet() ) {
+          System.err.println("Test " + methodName + " failed under testMultithreadedAuthenticate:");
+          exceptions.get(methodName).printStackTrace();
+        }
+        throw exceptions.elements().nextElement();
+      }
     }
 }

